### PR TITLE
feat: JDownloader

### PR DESCRIPTION
**Pull request by Aethaeran**
## Fixes issues: 
- Having to install JDownloader 2 manually.
- It is a requested feature on swizzin's feathub. `https://feathub.com/liaralabs/swizzin/+79`

## Proposed Changes:
- Adding ".idea" folder to .gitignore for JetBrains IDE users.
- Adding install, remove and test scripts for JDownloader 2

## Change Categories

- New feature 
- Contribution flow 

## Checklist

- [x] Branch was made off the `develop` branch and the PR is targetting the `develop` branch

- [ ] Docs have been made OR are not necessary
    - PR link: 
- [ ] Changes to panel have been made OR are not necessary
    - PR link: 
- [x] Code is formatted [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [x] Code conforms to project structure [(See more)](https://swizzin.ltd/dev/structure)
- [x] Shellcheck isn't screaming [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [x] Prints to terminal are handled [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#printing-into-the-terminal)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] Testing was done
   - [ ] Tests created or no new tests necessary
   - [ ] Tests executed

### Key
✅ = Works successfully
❎ = Does not work BUT is handled gracefully
🛠 = Still WIP
❌ = Broken / not working
### Scenarios
- Fresh app install with correct JD information
    - ✅ With only one user 
    - 🛠 With multiple users present 
Currently this passes the details from the first user too. This isn't intended. Needs a fix.
- ✅ Fresh app install with incorrect JD information
    - ✅ With only one user 
    - ✅ With multiple users present
- Fresh app install with correct details for one user, and incorrect details for another user.
- 🛠 Bypass MyJDownloader detail passing entirely with environment variable
- Remove Install
    - With only one user
    - With multiple users present
- `box`  adduser\deluser adds\removes user installation. 
Simply handled by adding JDownloader to the lists of other apps already handled in a similar fashion in `box` itself.
    - `box adduser`
    - `box deluser`
- ❎ Update\Upgrade from version in master 
Deemed unnecessary due to JDownloader handling updates internally.
- ❎ Password gets changed from `box` in app 
This isn't possible. JDownloader2's Web-UI is intended to be used through their my.jdownloader.org service. The odds that the swizzin user data is going to align with your account details on their site is unlikely. The username for my.jdownloader.org is always in the form of an e-mail. Furthermore, JDownloader saves your my.jdownloader.org details in plain text in the cfg directory. We wouldn't want to have unhashed swizzin passwords just lying around.
image
### Architectures Tests

|   			| `amd64` 	| `armhf` 	| `arm64` 	| Unspecified 	|
|--------		|-------- 	|-------- 	|-------- 	|----------		|
| Focal 		|	🛠		|	🛠		|	🛠		|	🛠			|
| Bionic		|	🛠		|	🛠		|	🛠		|	🛠			|
| Buster		|	🛠		|	🛠		|	🛠		|	🛠			|
| Stretch		|	🛠		|	🛠		|	🛠		|	🛠			|
| Raspbian  	|	🛠		|	🛠		|	🛠		|	🛠			|

### CODE REVIEW

File: scripts/install/jdownloader.sh [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+# https://swizzin.ltd/dev/structure/
+# https://www.digitalocean.com/community/tutorials/how-to-install-java-with-apt-on-debian-10
+# https://support.jdownloader.org/Knowledgebase/Article/View/install-jdownloader-on-nas-and-embedded-devices
+# Liara already made a doc for installing JDownloader manually https://docs.swizzin.net/guides/jdownloader/
+# https://linuxize.com/post/how-to-check-if-string-contains-substring-in-bash/
+# https://linuxize.com/post/bash-check-if-file-exists/
+
+# TODO: Figure out if these imports are even necessary
+. /etc/swizzin/sources/globals.sh
+. /etc/swizzin/sources/functions/users
```
Comment (flying-sausages) [inline scripts/install/jdownloader.sh:15]: Not necessary, these get pulled in whenever `box` is ran and are exported into the shell for the whole `box` run

File: scripts/install/jdownloader.sh [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+# Functions
+function install_jdownloader() {
+
+    # Get my.jdownloader info for user
+    # TODO: Should double check to confirm everything is accurate, and loop back if anything isn't filled out.
+    # TODO: swizzin likely has utils for this already
+    echo_query "The user will require an account at https://my.jdownloader.org/ in order to access your JDownloader installation's web UI.\nIt is recommended to use a randomly generated password for the account since the password is save in plain text in the program folder.\nEnter the e-mail used to access this account once one is created:"
```
Comment (flying-sausages) [inline scripts/install/jdownloader.sh:23]: I'd break this up a bit
```suggestion
    echo_info "The user will require an account at https://my.jdownloader.org/ in order to access your JDownloader installation's web UI.\nIt is recommended to use a randomly generated password for the account since the password is save in plain text in the program folder."
    echo_query "Enter the e-mail used to access this account once one is created:"
```

File: scripts/install/jdownloader.sh [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    read -r 'myjd_email'
+    echo_query "Please enter the password for the account."
+    read -r 'myjd_password'
+    echo_query "Please enter the device name they would like this JDownloader installation to show up as."
```
Comment (flying-sausages) [inline scripts/install/jdownloader.sh:27]:
```suggestion
    echo_query "Please enter the desired device name"
```

File: scripts/install/jdownloader.sh [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    read -r 'myjd_devicename'
+
+    # Install JDownloader
+    echo_progress_start "Downloading and installing JDownloader for $user"
+    mkdir -p /home/"$user"/jd
+    wget -q http://installer.jdownloader.org/JDownloader.jar -O /home/"$user"/jd/JDownloader.jar
```
Comment (flying-sausages) [inline scripts/install/jdownloader.sh:33]: Do we need a whole `jar` for each user? why not use one `jar` and have all the services refer to that bin?

File: scripts/install/jdownloader.sh [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    SUB='not found'
+    if [[ "$STR" == *"$SUB"* ]]; then
+        echo_info "Java was not installed correctly. Check the swizzin log for details. Exiting."
+        exit 1
+    else
+        echo_info "Java was installed successfully."
+    fi
+else
+    echo_info "Java was found. No need to install."
+fi
```
Comment (flying-sausages) [inline scripts/install/jdownloader.sh:88]: use the `java` functions we already have in `sources/functions/java`, I think they're merged in with airsonic

File: scripts/install/jdownloader.sh [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+Environment=JD_HOME=/home/%i/jd
+ExecStart=/usr/bin/java -Djava.awt.headless=true -jar /home/%i/jd/JDownloader.jar
+
+[Install]
+WantedBy=multi-user.target
+EOF
+
+# Check for all current swizzin users, and install JDownloader for each user.
+users=("$(_get_user_list)")
+
```
Comment (flying-sausages) [inline scripts/install/jdownloader.sh:109]: Missing a closing `echo_progress_done` here to match the start from L91

Comment (flying-sausages) [reply scripts/install/jdownloader.sh:109]: The for loop call on L111 will make all sorts of prints, `starts` need to be directly closed by `done`

File: scripts/tests/jdownloader.sh [hunk 7]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+# TODO
```
Comment (flying-sausages) [inline scripts/tests/jdownloader.sh:2]: If you delete this file and `box test jdownloader` runs just fine after an install, you don't even need this. The default tests can go far, test files are only needed for exceptions to the defaults.

**Review by flying-sausages — COMMENTED** [review]
Quick read through while I'm on vacay, I'll give this a shot sometime at the end of the week maybe

File: scripts/install/jdownloader.sh (hunk 3 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh:27]: Yeah I get way too wordy. Acceptable change in my books

File: scripts/install/jdownloader.sh (hunk 2 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh:23]: Also a good call. Honestly still haven't looked at the differences between the different swizzin echos. But this another time where I get a bit wordy. I'm sure this could be condensed further still.

File: scripts/install/jdownloader.sh (hunk 1 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh:15]: I had a feeling if I left a TODO about that someone would chime in for me. Appreciated!

File: scripts/install/jdownloader.sh [hunk 8]
```diff
@@ -20,11 +20,12 @@ function install_jdownloader() {
     # Get my.jdownloader info for user
     # TODO: Should double check to confirm everything is accurate, and loop back if anything isn't filled out.
     # TODO: swizzin likely has utils for this already
-    echo_query "The user will require an account at https://my.jdownloader.org/ in order to access your JDownloader installation's web UI.\nIt is recommended to use a randomly generated password for the account since the password is save in plain text in the program folder.\nEnter the e-mail used to access this account once one is created:"
+    echo_info "The user will require an account at https://my.jdownloader.org/ in order to access your JDownloader installation's web UI.\nIt is recommended to use a randomly generated password for the account since the password is save in plain text in the program folder."
```
Comment (flying-sausages) [inline scripts/install/jdownloader.sh:23]:
```suggestion
    echo_info "An account from https://my.jdownloader.org/ is required in order to access the web UI.\nUse a randomly generated password at registration as the password is stored in plain text."
```

Comment (Aethaeran) [reply scripts/install/jdownloader.sh]: Thank you!

File: scripts/install/jdownloader.sh (hunk 4 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh:34]: b said something similar on the Discord. I will certainly be looking into this once everything else is said and done. I'm not 100% sure which files can be duplicated across installations. Although we would certainly need the `cfg` directory generated by each intallation to be separate. As there is an absolute plethora of options for JDownloader, and at the very minimum each users account information for accessing the web ui via the third party would have to be self contained.

It's quite possible that I could do something like drop the JDownloader.jar into /opt/jdownloader instead of to each user's home directory. The service file passes the working directory for JDownloader, so where the actual .jar is containted could be inconsequential.

However, I'm also not completely certain that that JDownloader.jar file will not be replaced at some point by an internal update. My concern here being that if one user's configuration was not updated along with the application files (for instance if they had jdownloader disabled for some time) then we might end up with some broken installations.

TL;DR - I'm not sure that having multiple copies of a 3.8 MB .jar file is something we should be worried about if it might cause more headaches down the line.

File: scripts/install/jdownloader.sh (hunk 5 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh:89]: We did discuss this in passing on the Discord. I believe we decided to just have me do my own thing? As the java scripts for airsonic were Java 8 specific for some reason or another. I don't think JDownloader will care either way, because as far as I know it only requires `Java 1.8`.

I'm open to discuss it further if you would like.

File: scripts/tests/jdownloader.sh (hunk 7 above)
Comment (Aethaeran) [reply scripts/tests/jdownloader.sh:2]: I did not realize that `box test jdownloader` would do anything without this file. I will have to look into the default operations of `box test`.

However, I am thinking I would like to add a custom test. Specifically, I'm thinking something to detect if their installation is accessible via the third party website. Since the login information for that will be stored locally in plain text anyways. This would give sysadmins a quick test to see if any of the login information was entered incorrectly on install, if their login data has changed during install, or if the installation is unaccessible.

File: scripts/install/jdownloader.sh (hunk 6 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh]: I think I resolved this since my changes today. You'll have to take another look and let me know.

File: scripts/install/jdownloader.sh (hunk 4 above)
Comment (flying-sausages) [reply scripts/install/jdownloader.sh]: Ait, let's keep this thread open until the end of the PR and let's revisit it then

File: scripts/install/jdownloader.sh (hunk 5 above)
Comment (flying-sausages) [reply scripts/install/jdownloader.sh]: I don't see a reason to deviate from JRE8 and using the wheel we already invented. Maintaining an extra install of JRE as we'll have to start maintaining those in all of the service files for Java apps.

Is there any outstanding advantage when using +8, which would outweigh the added complexity to the system?

File: scripts/tests/jdownloader.sh (hunk 7 above)
Comment (flying-sausages) [reply scripts/tests/jdownloader.sh:2]: Fair point that I should write docs for the testing functionality.

I would say that we should not rely on using tools from systems outside of our control, if that portal goes down we need a rewrite on our end. As long as the services are up, and they're accessible internally and via nginx, we're good I would say.

**Comment by flying-sausages** [comment]
image
FYI

File: scripts/install/jdownloader.sh (hunk 5 above)
Comment (flying-sausages) [reply scripts/install/jdownloader.sh]: I think it would actually be nice to move some of the "testing" logic into the function defined here so that we use it anytime we install JDK

https://github.com/swizzin/swizzin/blob/master/sources/functions/java

this means we should handle the entire java install above hrough something like
```sh
. /etc/swizzin/sources/functions/java
install_jdk8
```

File: scripts/tests/jdownloader.sh (hunk 7 above)
Comment (Aethaeran) [reply scripts/tests/jdownloader.sh:2]: I would rather access it directly as well. I'm not actually sure there is an alternative though. As far as I know JDownloader's web UI can **only** be used via their site. I thought I might have seen something contrary to this in passing the other day though... I'll have to take another look around.

Comment (flying-sausages) [reply scripts/tests/jdownloader.sh:2]: We can just check if the ports are open and being listened to, not necessarily make some `curl`/HTTP requests to them. We have a basetest function for this. I would say that's _good enough_

Comment (Aethaeran) [reply scripts/tests/jdownloader.sh:2]: Oh alright then. That does sound like a better solution to me. Where might I find this function?

File: scripts/install/jdownloader.sh (hunk 5 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh]: This change seems fine to me. Like I said, I don't think JDownloader is going to are either way.

Except, I should note that I've moved away from:
```
STR=$(java --version)
SUB='not found'
if [[ "$STR" == *"$SUB"* ]]; then
```
because it was not functioning properly under certain conditions. I'm not sure I could replicate them but it was because `java --version` was outputting something other than "command not found" when Java was not installed, and therefore skipping installation. I think possibly the system had just recently had Java removed, but I could be mistaken.
I have been using this instead since. Haven't had any issues since. 
```
if [[ ! -e /usr/bin/java ]]; then
```
I'm not sure where you would want to put them into the `install_java8` function itself. I'm a little confused because it looks like this function only handles `buster` installs. I'm I reading this incorrectly?

File: scripts/tests/jdownloader.sh (hunk 7 above)
Comment (flying-sausages) [reply scripts/tests/jdownloader.sh:2]: Check this https://github.com/swizzin/swizzin/blob/master/sources/functions/tests, let me know if the inline docs aren't good enough

File: scripts/install/jdownloader.sh (hunk 5 above)
Comment (flying-sausages) [reply scripts/install/jdownloader.sh]: > I'm not sure where you would want to put this logic into the install_java8 function itself. I'm a little confused because it looks like this function only handles buster installs. Am I reading this incorrectly?

It handles buster explicitly, and all others (the `*)` case) are handled through the apt install.

We _might_ want to check for **pre-existing** JRE installs before the actual install here and try to exit gracefully or offer some mitigations.

File: scripts/tests/jdownloader.sh (hunk 7 above)
Comment (flying-sausages) [reply scripts/tests/jdownloader.sh:2]: ++ https://swizzin.ltd/dev/structure/#test

File: scripts/install/jdownloader.sh (hunk 5 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh]: Ahhhhh see. I didn't know about `*)` case. This makes more sense now. If I remember correctly. You mentioned something about airsonic requiring Java 8 specifically for some reason or another. So for the install_java8 function, I'd recommend checking for `/usr/bin/java` first, and then check to see if a version other than 8 is installed. Then have some sort of user intervention at that point. They could choose to (A) Cancel. (B) Remove their current installation for a Java 8 install instead. (C) If this is somehow possible, opt into having parallel java installations. (All apps that _require_ `Java 8` would have to act accordingly)

Comment (Aethaeran) [reply scripts/install/jdownloader.sh]: This is very rough, but essentially something like this.
```
if [[ ! -e /usr/bin/java ]]; then
    echo_info "Java is not installed. We are free to install Java 8"
else
    echo_info "Java is installed. Check which version"
    if [[ ! "$(/usr/bin/java --version)" == *'build 8.'* ]]; then
        echo_error "Current Java version is not 8."
        echo_info "We can handle this in three ways:"
        echo_info "(A) "
        echo_info "(B) "
        echo_info "(C) "
        echo_query "How do you want to handle this?"
        read -r "answer"
        case $answer in
            "A")
                echo_info "Do A"
                ;;
            "B")
                echo_info "Do B"
                ;;
            "C")
                echo_info "Do C"
                ;;
            *)
                echo_info "Loop back and get a proper answer."
                ;;
        esac
    fi
fi
```

**Comment by Aethaeran** [comment]
> 
> 
> image
> FYI

I've been pushing my scripts from a Windows installation. They all come up 755 when I check them with git bash. I'm assuming it's being lost on push since Windows doesn't handle perms this way.

File: scripts/install/jdownloader.sh (hunk 5 above)
Comment (flying-sausages) [reply scripts/install/jdownloader.sh]: Hm.

Fuck it, that's for another PR. We should re-evaluate if it's mybe time to upgrade the JRE too, but that's out of scope for this app introduction. If we use the function we have for it now, then making the changes there later will apply to the whole toolset, and that will be slick.

Comment (Aethaeran) [reply scripts/install/jdownloader.sh]: Sounds good to me. Will change it as soon as I have all the install issues sorted out

File: scripts/tests/jdownloader.sh (hunk 7 above)
Comment (Aethaeran) [reply scripts/tests/jdownloader.sh]: Thinking the default tests aren't going to cut it in this case.
```
# box test jdownloader
INFO	Running default test for jdownloader.
	This test is likely to fail in case the item is not a 
	standard service+nginx app
...	Checking jdownloader service is active
Unit jdownloader.service could not be found.
WARN	jdownloader.service not active!
...	Checking if port jdownloader is open and listened to
WARN	Couldn't guess port
...	Checking if port jdownloader is reachable via curl
WARN	Couldn't guess port
WARN	nginx not installed, cannot verify if jdownloader is 
	reachable through it. Skipping.
ERROR	Errors were encountered
	Please consult the above and/or check the log (less -R +G 
	/root/logs/tests-2021.07.30-08:25:13.txt)
```

Comment (Aethaeran) [reply scripts/tests/jdownloader.sh]: BUT... 
I did find these today, and they suggest that it might be possible to access JDownloader without using the third party website.
[A help topic suggesting that NGINX might be an option.](https://board.jdownloader.org/showthread.php?t=76156)
[API Docs for reference](https://my.jdownloader.org/developers/)

File: scripts/install/jdownloader.sh (hunk 4 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh]: Upon further review, it would seem that JDownloader really doesn't care where JD_HOME is set to (during the intial runs at least). It seems to generate the files around itself, wherever it may be being run from.

Don't have high hopes of this being an option at this point.

File: scripts/install/jdownloader.sh (hunk 5 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh]: The script now installs Java with the swizzin function.
What should we call this other PR for Java changes? Should I start it? or should you?

File: scripts/install/jdownloader.sh (hunk 6 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh]: I've added an awful lot of echos since this comment. Might need an additional review to see if they are excessive.

File: scripts/install/jdownloader.sh (hunk 4 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh]: Just thought I'd give it a brief Google search.
[Doesn't sound promising.](https://board.jdownloader.org/showthread.php?t=74492)

File: scripts/tests/jdownloader.sh (hunk 7 above)
Comment (flying-sausages) [reply scripts/tests/jdownloader.sh]: Yeah checking the web UI might be more trouble than it's worth honestly.

File: scripts/install/jdownloader.sh (hunk 5 above)
Comment (flying-sausages) [reply scripts/install/jdownloader.sh]: I'd say `fix: java management` is good enough. Feel free to have a start at it ;)

File: scripts/install/jdownloader.sh (hunk 4 above)
Comment (flying-sausages) [reply scripts/install/jdownloader.sh]: Fair, that pretty much says it's not possible and supported so I'm closing this one.

File: scripts/install/jdownloader.sh (hunk 6 above)
Comment (flying-sausages) [reply scripts/install/jdownloader.sh]: I cleaned up a majority of the echos

**Comment by flying-sausages** [comment]
This is what the echos look like right now. Looks like we still need to fix a couple bugs about the intialisation sequence.

image

This is what the service dies on when started for my user
```
Jul 31 17:53:03 m1 java[88140]: 23|Log.L.log 7/31/21 3:53:03 PM - SEVERE [ Log.L.log ] -> Exception thrown at jd.SecondLevelLaunch$9$1.run(SecondLevelLaunch.java:919):
Jul 31 17:53:03 m1 java[88140]: java.lang.ExceptionInInitializerError
Jul 31 17:53:03 m1 java[88140]:         at jd.SecondLevelLaunch$9$1.run(SecondLevelLaunch.java:747)
Jul 31 17:53:03 m1 java[88140]: Caused by: java.lang.RuntimeException: No Console Available!
Jul 31 17:53:03 m1 java[88140]:         at org.appwork.console.ConsoleDialog.<init>(ConsoleDialog.java:54)
Jul 31 17:53:03 m1 java[88140]:         at org.appwork.console.ConsoleDialog.<init>(ConsoleDialog.java:60)
Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.askLoginsOnConsole(MyJDownloaderController.java:273)
Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.start(MyJDownloaderController.java:125)
Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.<init>(MyJDownloaderController.java:112)
Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.<clinit>(MyJDownloaderController.java:37)
Jul 31 17:53:03 m1 java[88140]:         ... 1 more
```

File: scripts/tests/jdownloader.sh (hunk 7 above)
Comment (Aethaeran) [reply scripts/tests/jdownloader.sh]: You might not be wrong there. It's also entirely possible that the JDownloader logs would be telling us if it failed anyways.

**Comment by Aethaeran** [comment]
> 
> 
> This is what the echos look like right now. Looks like we still need to fix a couple bugs about the intialisation sequence.
> 
> image
> 
> This is what the service dies on when started for my user
> 
> ```
> Jul 31 17:53:03 m1 java[88140]: 23|Log.L.log 7/31/21 3:53:03 PM - SEVERE [ Log.L.log ] -> Exception thrown at jd.SecondLevelLaunch$9$1.run(SecondLevelLaunch.java:919):
> Jul 31 17:53:03 m1 java[88140]: java.lang.ExceptionInInitializerError
> Jul 31 17:53:03 m1 java[88140]:         at jd.SecondLevelLaunch$9$1.run(SecondLevelLaunch.java:747)
> Jul 31 17:53:03 m1 java[88140]: Caused by: java.lang.RuntimeException: No Console Available!
> Jul 31 17:53:03 m1 java[88140]:         at org.appwork.console.ConsoleDialog.<init>(ConsoleDialog.java:54)
> Jul 31 17:53:03 m1 java[88140]:         at org.appwork.console.ConsoleDialog.<init>(ConsoleDialog.java:60)
> Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.askLoginsOnConsole(MyJDownloaderController.java:273)
> Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.start(MyJDownloaderController.java:125)
> Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.<init>(MyJDownloaderController.java:112)
> Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.<clinit>(MyJDownloaderController.java:37)
> Jul 31 17:53:03 m1 java[88140]:         ... 1 more
> ```
Just did some clean-up to the logic that should clear up the grep spam.

image

Are you not passing authentic details to the MY JD prompts? I believe that is the only reason you can get the the "No Console Available" error. That's because the headless JDownloader will try to prompt the user for the MYJD details. Since it is being run in the background, instead of a regular terminal, there is no console to request this information in, so it throws this error. I'm hoping ultimately we can use this to verify MYJD details.

Does it still say it's active like this?

image

Because when I put in accurate details I stall out somewhere around here

image

**Comment by brettpetch** [comment]
> > This is what the echos look like right now. Looks like we still need to fix a couple bugs about the intialisation sequence.
> > image
> > This is what the service dies on when started for my user
> > ```
> > Jul 31 17:53:03 m1 java[88140]: 23|Log.L.log 7/31/21 3:53:03 PM - SEVERE [ Log.L.log ] -> Exception thrown at jd.SecondLevelLaunch$9$1.run(SecondLevelLaunch.java:919):
> > Jul 31 17:53:03 m1 java[88140]: java.lang.ExceptionInInitializerError
> > Jul 31 17:53:03 m1 java[88140]:         at jd.SecondLevelLaunch$9$1.run(SecondLevelLaunch.java:747)
> > Jul 31 17:53:03 m1 java[88140]: Caused by: java.lang.RuntimeException: No Console Available!
> > Jul 31 17:53:03 m1 java[88140]:         at org.appwork.console.ConsoleDialog.<init>(ConsoleDialog.java:54)
> > Jul 31 17:53:03 m1 java[88140]:         at org.appwork.console.ConsoleDialog.<init>(ConsoleDialog.java:60)
> > Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.askLoginsOnConsole(MyJDownloaderController.java:273)
> > Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.start(MyJDownloaderController.java:125)
> > Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.<init>(MyJDownloaderController.java:112)
> > Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.<clinit>(MyJDownloaderController.java:37)
> > Jul 31 17:53:03 m1 java[88140]:         ... 1 more
> > ```
> 
> Just did some clean-up to the logic that should clear up the grep spam.
> 
> image
> 
> Are you not passing authentic details to the MY JD prompts? I believe that is the only reason you can get the the "No Console Available" error. That's because the headless JDownloader will try to prompt the user for the MYJD details. Since it is being run in the background, instead of a regular terminal, there is no console to request this information in, so it throws this error. I'm hoping ultimately we can use this to verify MYJD details.
> 
> Does it still say it's active like this?
> 
> image
> 
> Because when I put in accurate details I stall out somewhere around here
> 
> image

I’d suggest use the default masteruser username and password for this. You should be able to find the _get_user_password function pretty easily.

**Comment by Aethaeran** [comment]
> > > This is what the echos look like right now. Looks like we still need to fix a couple bugs about the intialisation sequence.
> > > image
> > > This is what the service dies on when started for my user
> > > ```
> > > Jul 31 17:53:03 m1 java[88140]: 23|Log.L.log 7/31/21 3:53:03 PM - SEVERE [ Log.L.log ] -> Exception thrown at jd.SecondLevelLaunch$9$1.run(SecondLevelLaunch.java:919):
> > > Jul 31 17:53:03 m1 java[88140]: java.lang.ExceptionInInitializerError
> > > Jul 31 17:53:03 m1 java[88140]:         at jd.SecondLevelLaunch$9$1.run(SecondLevelLaunch.java:747)
> > > Jul 31 17:53:03 m1 java[88140]: Caused by: java.lang.RuntimeException: No Console Available!
> > > Jul 31 17:53:03 m1 java[88140]:         at org.appwork.console.ConsoleDialog.<init>(ConsoleDialog.java:54)
> > > Jul 31 17:53:03 m1 java[88140]:         at org.appwork.console.ConsoleDialog.<init>(ConsoleDialog.java:60)
> > > Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.askLoginsOnConsole(MyJDownloaderController.java:273)
> > > Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.start(MyJDownloaderController.java:125)
> > > Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.<init>(MyJDownloaderController.java:112)
> > > Jul 31 17:53:03 m1 java[88140]:         at org.jdownloader.api.myjdownloader.MyJDownloaderController.<clinit>(MyJDownloaderController.java:37)
> > > Jul 31 17:53:03 m1 java[88140]:         ... 1 more
> > > ```
> > 
> > Just did some clean-up to the logic that should clear up the grep spam.
> > 
> > image
> > 
> > Are you not passing authentic details to the MY JD prompts? I believe that is the only reason you can get the the "No Console Available" error. That's because the headless JDownloader will try to prompt the user for the MYJD details. Since it is being run in the background, instead of a regular terminal, there is no console to request this information in, so it throws this error. I'm hoping ultimately we can use this to verify MYJD details.
> > 
> > Does it still say it's active like this?
> > 
> > image
> > 
> > Because when I put in accurate details I stall out somewhere around here
> > 
> > image
> 
> I’d suggest use the default masteruser username and password for this. You should be able to find the _get_user_password function pretty easily. 

Unfortunately the swizzin user and password are never going to line up with the my.jdownloader.org details. I have now added in logic to verify the my jdownloader details are correct. Later on, I will add in a little bit more logic later to bypass inputting my.jdownloader.org account details.

Give it another spin and let me know how it goes. Interested to hear how it goes.

File: scripts/install/jdownloader.sh [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    fi
+
+    cat > "$JD_HOME/cfg/org.jdownloader.api.myjdownloader.MyJDownloaderSettings.json" << EOF
+{
+    "email" : "$MYJD_EMAIL",
+    "password" : "$MYJD_PASSWORD",
+    "devicename" : "$MYJD_DEVICENAME"
+}
+EOF
+}
```
Comment (brettpetch) [inline scripts/install/jdownloader.sh:54]: this *could* be something in a functions file.

File: scripts/remove/jdownloader.sh [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+. /etc/swizzin/sources/functions/utils
+
+# this will check for all swizzin users, and iterate and remove all JDownloader installations.
+users=("$(_get_user_list)")
+# The following line cannot follow SC2068 because it will cause the list to become a string.
+# shellcheck disable=SC2068
+for user in ${users[@]}; do
+    echo_progress_start "Removing JDownloader for $user..."
+    systemctl disable -q --now jdownloader@"$user"
+    # TODO: Do a check to see if they want to purge their JDownloader configurations.
```
Comment (brettpetch) [inline scripts/remove/jdownloader.sh:16]: we don't currently have purge functionality built into swizzin. It is being looked at for our v4 milestone as a box flag. If you did want to include this option, it might be a good idea to push into its own purge function ;)

File: scripts/install/jdownloader.sh (hunk 9 above)
Comment (Aethaeran) [reply scripts/install/jdownloader.sh:54]: Will likely be. Going to add a catch to inject MyJDownloader details into installations that don't have proper details yet.

Might have to move some of the MyJD verification into this function as well.

File: scripts/remove/jdownloader.sh (hunk 10 above)
Comment (Aethaeran) [reply scripts/remove/jdownloader.sh:16]: Was assuming it would be purged on removal, or that I could add a simple yes/no here.
deluser purges all configurations, so I guess I just assumed a feature removal would as well.

Comment (Aethaeran) [reply scripts/remove/jdownloader.sh]: I've now added the yes/no prompts for configuration purging on removal.

At the very least I'd like to leave it this way until testing is thoroughly completed, then just perhaps move that logic and operation to a purge_jdownloader function.

File: scripts/install/jdownloader.sh [hunk 11]
```diff
@@ -182,11 +182,12 @@ EOF
 _systemd
 
 # TODO: Add environment variable to bypass the following block. For unattended installs.
-if ask "Do you want to add ANY MyJDownloader account information for users?\nIt is required for them to access the web UI." N; then
-    NO_MYJD="false" # If no
-else
-    NO_MYJD="true" # If yes
-fi
+if [[ -n "${BYPASS_MYJDOWNLOADER}" ]]; then
```
Comment (flying-sausages) [inline scripts/install/jdownloader.sh:185]: Let's stick to the `MYJD_` prefix please, it will be a lot more coherent in an env file down the road
